src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+			/// </summary>
+			/// <param name="enlistment">The enlistment object for signaling to the transaction manager the notification has been handled.</param>
+			/// <param name="success"><see langword="true"/> if this is a commit callback, <see langword="false"/> if this is a rollback
+			/// callback, <see langword="null"/> if this is an in-doubt callback.</param>
+			protected virtual void ProcessSecondPhase(Enlistment enlistment, bool? success)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					_logger.Debug(
+						success.HasValue
+							? success.Value
+								? "Committing system transaction"
+								: "Rolled back system transaction"
+							: "System transaction is in doubt");
+					// we have not much to do here, since it is the actual
+					// DB connection that will commit/rollback the transaction
+					// After transaction actions are raised from TransactionCompleted event.
+
+					enlistment.Done();
+				}
+			}
+
+			#endregion
+
+			/// <summary>
+			/// Handles the transaction completion event. Notify <see cref="ConnectionManager"/> of the end of the
+			/// transaction. Notifies end of transaction to the session and to <see cref="ConnectionManager.DependentSessions"/>
+			/// if any. Close sessions requiring it then cleanup transaction contextes and then <see cref="Unlock"/> blocked
+			///  threads.
+			/// </summary>
+			/// <param name="sender">The object issuing the event.</param>
+			/// <param name="e">The event argument.</param>
+			protected virtual void TransactionCompleted(object sender, TransactionEventArgs e)
+			{
+				try
+				{
+					e.Transaction.TransactionCompleted -= TransactionCompleted;
+					// This event may execute before second phase, so we cannot try to get the success from second phase.
+					// Using this event is required by example in case the prepare phase failed and called force rollback:
+					// no second phase would occur for this ressource. Maybe this may happen in some other circumstances
+					// too.
+					var wasSuccessful = false;
+					try
+					{
+						wasSuccessful =
+							AmbientTransaction.TransactionInformation.Status == TransactionStatus.Committed;
+					}
+					catch (ObjectDisposedException ode)
+					{
+						_logger.Warn("Completed transaction was disposed, assuming transaction rollback", ode);
+					}
+
+					// Allow transaction completion actions to run while others stay blocked.
+					_bypassWait.Value = true;
+					using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+					{
+						// Flag active as false before running actions, otherwise the session may not cleanup as much
+						// as possible.
+						IsInActiveTransaction = false;
+						// Never allows using connection on after transaction event. And tell the connection manager
+						// it is called from system transaction.
+						using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(false))
+						{
```

Comment:
Now inside `using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(false))` for:

1. forbidding connection usage (`false` parameter), which by the way never causes a connection switch.
2. just flagging the connection as requiring a release, instead of releasing it from transaction completion event (NH-3023).